test/cpp/microbenchmarks/callback_unary_ping_pong.h

```diff
@@ -0,0 +1,101 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Benchmark gRPC end2end in various configurations */
+
+#ifndef TEST_CPP_MICROBENCHMARKS_CALLBACK_UNARY_PING_PONG_H
+#define TEST_CPP_MICROBENCHMARKS_CALLBACK_UNARY_PING_PONG_H
+
+#include <benchmark/benchmark.h>
+#include <sstream>
+#include "src/core/lib/profiling/timers.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/cpp/microbenchmarks/callback_test_service.h"
+#include "test/cpp/microbenchmarks/fullstack_context_mutators.h"
+#include "test/cpp/microbenchmarks/fullstack_fixtures.h"
+
+namespace grpc {
+namespace testing {
+
+/*******************************************************************************
+ * BENCHMARKING KERNELS
+ */
+
+// Send next rpc when callback function is evoked.
+void SendCallbackUnaryPingPong(benchmark::State& state, EchoRequest* request,
+                               EchoResponse* response,
+                               EchoTestService::Stub* stub_, bool& done,
+                               std::mutex& mu, std::condition_variable& cv) {
+  int response_msgs_size = state.range(1);
+  ClientContext* cli_ctx = new ClientContext();
```

Comment:
The reason why I allocate client context to heap is because callback rpc call needs a client context struct for callback result reference. If I use stack here, context will be gone when SendCallbackUnaryPingPong is finished and callback hasn't evoked yet, which will lead to a segment fault.

Reply:
This is a reasonable point. The actual ClientContext object can be stack-allocated in the top-level microbenchmark function that calls this and then you can have an explicit destruction and placement new to recreate it before actually starting a new RPC. Something like

```
cli_ctx->~ClientContext();
new (cli_ctx) ClientContext();
```
before starting the new RPC. This works here since there is just 1 RPC outstanding at a time in these microbenchmarks.